File: src/main/java/dev/sky_lock/menu/InventoryMenuListener.java
package dev.sky_lock.menu;

import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.*;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.InventoryHolder;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.Optional;

/**
 * @author sky_lock
 */

public class InventoryMenuListener implements Listener {
    private final JavaPlugin plugin;

    public InventoryMenuListener(JavaPlugin plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        InventoryHolder holder = event.getInventory().getHolder();
        if (holder == null) {
            return;
        }
        if (!(holder instanceof InventoryMenu)) {
            return;
        }
        if (event.getSlotType() == InventoryType.SlotType.OUTSIDE) {
            return;
        }
        if (event.getClickedInventory().equals(event.getView().getBottomInventory())) {
            if (event.getClick().isShiftClick()) {
                event.setCancelled(true);
            }
            return;
        }
        if (!event.getInventory().equals(event.getClickedInventory())) {
            return;
        }
        if (event.getAction() == InventoryAction.HOTBAR_SWAP || event.getAction() == InventoryAction.HOTBAR_MOVE_AND_READD) {
            event.setCancelled(true);
            return;
        }
        if (!event.getClick().isRightClick() && !event.getClick().isLeftClick()) {
            event.setCancelled(true);
            return;
        }
        event.setCancelled(true);
        ((InventoryMenu) holder).click(event);
    }

    @EventHandler
    public void onInventoryDrag(InventoryDragEvent event) {
        InventoryMenu.of((Player) event.getWhoClicked()).ifPresent(menu -> {
            if (event.getInventory().equals(event.getView().getBottomInventory())) {
                if (event.getType() == DragType.SINGLE) {
                    return;
                }
            }
            event.setCancelled(true);
        });
    }

    @EventHandler
    public void onInventoryClose(InventoryCloseEvent event) {
        Optional<InventoryMenu> menu = InventoryMenu.of((Player) event.getPlayer());
        new BukkitRunnable() {
            @Override
            public void run() {
                menu.ifPresent(menu -> {
                    menu.remove((Player) event.getPlayer());
                });
            }
        }.runTaskLater(plugin, 1L);
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        InventoryMenu.of(event.getPlayer()).ifPresent(menu -> menu.close(event.getPlayer()));
    }

}

File: src/main/kotlin/dev/sky_lock/menu/InventoryMenuListener.kt
package dev.sky_lock.menu

import dev.sky_lock.menu.InventoryMenu.Companion.of
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.*
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitRunnable

/**
 * @author sky_lock
 */
class InventoryMenuListener(private val plugin: JavaPlugin) : Listener {
    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        val holder = (event.inventory.holder ?: return) as? InventoryMenu ?: return
        if (event.slotType == InventoryType.SlotType.OUTSIDE) {
            return
        }
        if (event.clickedInventory == event.view.bottomInventory) {
            if (event.click.isShiftClick) {
                event.isCancelled = true
            }
            return
        }
        if (event.inventory != event.clickedInventory) {
            return
        }
        if (event.action == InventoryAction.HOTBAR_SWAP || event.action == InventoryAction.HOTBAR_MOVE_AND_READD) {
            event.isCancelled = true
            return
        }
        if (!event.click.isRightClick && !event.click.isLeftClick) {
            event.isCancelled = true
            return
        }
        event.isCancelled = true
        holder.click(event)
    }

    @EventHandler
    fun onInventoryDrag(event: InventoryDragEvent) {
        of((event.whoClicked as Player)).ifPresent { menu: InventoryMenu? ->
            if (event.inventory == event.view.bottomInventory) {
                if (event.type == DragType.SINGLE) {
                    return@ifPresent
                }
            }
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onInventoryClose(event: InventoryCloseEvent) {
        val menu = of((event.player as Player))
        object : BukkitRunnable() {
            override fun run() {
                menu.ifPresent { menu: InventoryMenu -> menu.remove((event.player as Player)) }
            }
        }.runTaskLater(plugin, 1L)
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        of(event.player).ifPresent { menu: InventoryMenu -> menu.close(event.player) }
    }

}